L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,501 @@
+// If there is no certificate provider, the system default root certificates will be used to
+// verify server certificates. 
+// If a certificate provider is set and it provides root certifices that root
+// will be used.
+// If a certificate provider is set and it provides identity credentials those
+// identity credentials will be used.
+class TlsChannelCredentialsOptions final : public TlsCredentialsOptions {
+ public:
+  // Sets the decision of whether to do a crypto check on the server certificates.
+  // The default is true.
+  void set_verify_server_certificates(bool verify_server_certs);
+}
+
+// Builds a ChannelCredentials instance that establishes TLS connections in the manner specified by options.
+std::shared_ptr<ChannelCredentials> TlsCredentials(
+    const TlsChannelCredentialsOptions& options);
+}
+
+// Builds a ServerCredentials instance that establishes TLS connections in the manner specified by options.
+std::shared_ptr<ServerCredentials> TlsServerCredentials(
+    const TlsServerCredentialsOptions& options);
+```
+
+To make migration from the `SslCredentials` stack to the `TlsCredentials` stack easier,
+we will also have a constructor for `TlsChannelCredentialsOptions` and
+`TlsServerCredentialsOptions` that accept their corresponding `SslOptions`
+structs.
+
+### Credential Reloading
+Credential reloading will be implemented via a `CertificateProviderInterface`.
+Note that there is a naming mismatch here - there exist credentials beyond
+certificates that this _could_ theoretically support.  This provider is
+responsible for sourcing key material and supplying it to the
+internal stack via the `CertificateProvider`'s `SetKeyMaterials` API.  Two
+reference implementations of the `CertificateProviderInterface` are provided -
+`StaticDataCertificateProvider` and `FileWatcherCertificateProvider`. The
+`StaticDataCertificateProvider` provides certificates from raw string data. The
+`FileWatcherCertificateProvider` will periodically and asynchronously refresh
+certificate data from a file, allowing changes to certificates without
+restarting the process. More details about these reference implementations
+are provided below. These implementations should cover many use cases.  If these
+providers do not support the intricacies for a specific use case, a user can
+provide their own implementations of the `CertificateProviderInterface`.
+
+```c
+/* Opaque types. */
+// A struct that stores the credential data presented to the peer in handshake
+// to show local identity. The private key and certificate chain must be PEM
+// encoded and the public key in the leaf certificate must correspond to the
+// given private key.
+struct IdentityKeyCertPair {
+  std::string private_key_pem;
+  std::string certificate_chain_pem;
+};
+
+
+/** ------------------------------------ Provider ------------------------------------ */
+/** Provides identity credentials and root certificates.
+*/
+class CertificateProviderInterface {
```

Comment:
I think we need to explain the "name" concept very carefully, because it's not obvious from the current text. At the top of this class (and the alternative version of the class), can we include something like the following:

"""
Provides credentials of several types, e.g. identity credentials or root certificates.

To enable the provider to manage multiple sets of credentials of each type, each set of credentials must be given a "name". The "name" is an opaque user-defined label, and the provider will reference this "name" e.g. in logs and error messages.
"""

Reply:
Added - I plan on adding more detail around this after we discuss the provider APIs today